src/qmapshack/mouse/CMouseSelect.cpp

```diff
@@ -26,6 +26,8 @@
 #include "mouse/CScrOptSelect.h"
 #include <QtWidgets>
 
+QMutex CMouseSelect::mutexPoisFound(QMutex::Recursive);
```

Comment:
@kiozen could you please have a look why this mutex isn't effective in stopping `draw()` being run parallel to `slotCopy()`? I'm pretty sure my segfaults are due to `poisFound` being updated while the copying is happening and can't figure out why this mutex doesn't lock.

Reply:
Because  `CMouseSelect::draw()` and `CMouseSelect::slotCopy()` are running in the same thread. See `void CCanvas::paintEvent(QPaintEvent*)` line 640. This is the main draw loop that resides in the main thread. The additional thread is triggered by `void IDrawContext::draw(QPainter& p, CCanvas::redraw_e needsRedraw, const QPointF& f)`.

Reply:
Maybe you get away with it by setting the `mutexPoisFound` to non-recursive and quit `findItems` when the mutex is locked. but I am not sure if that works in the same thread. Or you create a temporary real copy of the POI first. By that you can enter the eventloop and `findItems` can recreate the list without destroying your POI to copy.